Fix SliceDirectSelectiveStreamReader for dataLength is 0

yingsu00 opened the pull request:
For batch read mode, when dataLength is 0 and offsets array was not
reset, the offsets array for previous batch would be used to create
the output Block. This causes IndexOutOfBoundsException for downstream
operators. This fix initializes the offsets array to 0 for the batch
read mode.

```
== NO RELEASE NOTE ==
```

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java:
```diff
@@ -688,13 +688,13 @@ else if (isNotNull) {
             dataLength = totalLength;
             data = ensureCapacity(data, totalLength);
-            offsets = ensureCapacity(offsets, totalPositions + 1);
+            offsets = ensureCapacity(offsets, totalPositions + 1, SMALL, INITIALIZE);
```
I noticed that useBatchMode method has a lot of commented out code. Please, remove.

yingsu00 replied:
> I noticed that useBatchMode method has a lot of commented out code. Please, remove.

This is fixed in the second commit "Limit batch read mode to certain conditions for SliceDirectSelectiveStreamReader"

mbasmanova reviewed (commented):
"Fix SliceDirectSelectiveStreamReader for dataLength is 0" commit has at least 2 different changes, but only one is described in the commit message.

mbasmanova reviewed (commented):
I don't see any tests that cover these fixes. Please, add.

yingsu00 commented:
> I don't see any tests that cover these fixes. Please, add.

@mbasmanova thanks for reviewing. As stated in the PR message, I was unable to recreate the situation that causes the bug so far:

About testing: The test to reproduce this problem was hard to write. The SliceDirectSelectiveStreamReader needs to preserve the non-zero offsets array from the previous batches. We need two columns(INTEGER and VARCHAR), the first column has a TupleDomainFilter, and the second column doesn't have any filter. The second VARCHAR column would be using SliceDirectSelectiveStreamReader. For the second column(SliceDirectSelectiveStreamReader), the first batch needs to read data with dataLength > 0 and populate offsets array with non-zero values. In the second batch, the first column shall filter out enough rows so that the totalPositionCount for the second column is less than the previous batch and the offsets array has enough capacity and doesn't need to be allocated. Then the second column should have a null dataStream, and dataLength = 0 when read() is called, but there are still a few input positions. With the current test framework, it's hard to meet all these conditions therefore test is missing.

Do you have some idea how to meet these conditions? In general I feel we have very little control at the StreamReader level, and it's hard to reproduce many cases we saw in production using the current framework set up in `TestSelectiveOrcReader`. Is there any tests at the StreamReader level?

yingsu00 commented:
@mbasmanova How do you think if I introduce a new test `TestOrcSelectiveStreamReader`? In this test we can test the `read()` and `getBlock()` and other methods separately or in combination, and we can control StreamReader's internal state to simulate the corner cases we saw in production.

mbasmanova commented:
> @mbasmanova How do you think if I introduce a new test `TestOrcSelectiveStreamReader`? In this test we can test the `read()` and `getBlock()` and other methods separately or in combination, and we can control StreamReader's internal state to simulate the corner cases we saw in production.

Sounds good. Give it a try.

yingsu00 commented:
@mbasmanova I added `TestOrcSelectiveStreamReaders` and addressed your other comments. Will you be able to read again? Thank you very much and happy holiday season!

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -935,6 +928,19 @@ public static void assertFileContentsPresto(
+    static void assertBlockEquals(Type type, Block block, List<?> expectedValues, int offset)
+    {
+        int positionCount = block.getPositionCount();
+        List<Object> data = new ArrayList<>(positionCount);
```
This variable seems unnecessary. Why not just combine the two loops?

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceSelectiveStreamReader.java:
```diff
@@ -151,4 +152,10 @@ public static int computeTruncatedLength(Slice slice, int offset, int length, in
+    @VisibleForTesting
+    public void resetDataStream()
+    {
+        ((SliceDirectSelectiveStreamReader) directReader).resetDataStream();
```
shouldn't resetDataStream be called on dictionaryReader as well? or on currentReader?

yingsu00 replied:
@mbasmanova currentReader is only set after the first `startStripe()` is called. I added dictionaryReader too.

yingsu00 replied:
```
public class SliceSelectiveStreamReader
        implements SelectiveStreamReader
{
   @VisibleForTesting
    public void resetDataStream()
    {
        ((SliceDirectSelectiveStreamReader) directReader).resetDataStream();
        ((SliceDictionarySelectiveReader) dictionaryReader).resetDataStream();
    }
}
```

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/TestOrcSelectiveStreamReaders.java:
```diff
@@ -0,0 +1,116 @@
+        List<List<?>> values = ImmutableList.of(ImmutableList.of("a", ""));
+
+        TempFile tempFile = new TempFile();
+        writeOrcColumnsPresto(tempFile.getFile(), DWRF, CompressionKind.NONE, Optional.empty(), types, values, new OrcWriterStats());
```
Is this test specific to DWRF encoding? If not, perhaps, run it on all supported encodings.

yingsu00 replied:
Sure. I added the following settings
```
    private Set<OrcTester.Format> formats = ImmutableSet.of(ORC_12, ORC_11, DWRF);
    private Set<CompressionKind> compressions = ImmutableSet.of(NONE, SNAPPY, ZLIB, LZ4, ZSTD);
```

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/TestOrcSelectiveStreamReaders.java:
```diff
@@ -0,0 +1,116 @@
+     * The test issues two reads, the first one reads a non-empty string and populates non-zero offsets. The second one reads the empty string with the above conditions met.
+     */
+    @Test
+    public void testEmptySlice()
```
nit: perhaps, testAllEmptyStrings

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/TestOrcSelectiveStreamReaders.java:
```diff
@@ -0,0 +1,116 @@
+import static java.lang.Math.min;
+import static org.testng.Assert.assertEquals;
+
+public class TestOrcSelectiveStreamReaders
```
This name is very generic, but the test itself applies only to SliceDirectSelectiveStreamReader. Perhaps, rename to TestSliceDirectSelectiveStreamReader.

yingsu00 replied:
@mbasmanova I'm thinking to add more tests there, including more generic tests that cover all types, and also a test to resolve https://github.com/prestodb/presto/issues/15343 Do you think we can keep it as is for now?

mbasmanova replied:
Sounds good.

mbasmanova reviewed (commented):
@yingsu00 Thank you for the updates. Looks good to me % minor comments below.

yingsu00 commented:
@mbasmanova Masha, I updated the PR with the comments addressed and am waiting for the tests to pass. Do you have more comments? Thank you very much for reviewing!

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -935,6 +928,15 @@ public static void assertFileContentsPresto(
+    static void assertBlockEquals(Type type, Block block, List<?> expectedValues, int offset)
+    {
+        int positionCount = block.getPositionCount();
+        List<Object> data = new ArrayList<>(positionCount);
```
this variable is not used; remove

yingsu00 replied:
@mbasmanova Variable removed and tests passed. Thank you! Merry Christmas!

yingsu00 commented:
@mbasmanova I added one more fix for array of varchars and one fix for OrcTester. Will you be able to take a look? Thank you! cc @bhhari

mbasmanova reviewed (approved):
@yingsu00 Looks good to me. Please, run the verifier before merging.

yingsu00 commented:
Verifier run was already done https://www.internalfb.com/intern/presto/verifier/results/?test_id=48288
Also tested on the failed queries from 2020-12-26 to 2021-01-02 and they all succeeded.
